extensions/positron-python/python_files/posit/positron/positron_lsp.py

```diff
@@ -0,0 +1,1087 @@
+                error_handler=self.report_server_error,
+            )
+            logger.debug("LSP connection closed")
+            self._shutdown_server()
+
+        async def start_server() -> None:
+            # Use port=0 to let the OS pick a port
+            self._server = await asyncio.start_server(lsp_connection, host, 0)
+
+            # Find the port we're listening on
+            for socket in self._server.sockets:
+                addr, port = socket.getsockname()
+                if addr == host:
+                    logger.info("LSP server listening on %s:%d", host, port)
+                    break
+            else:
+                raise AssertionError("Unable to determine LSP server port")
+
+            # Notify the frontend that the server is ready
+            if self._comm is None:
+                logger.warning("LSP comm not set, cannot send server_started message")
+            else:
+                logger.info("LSP server ready, sending server_started message")
+                self._comm.send({"msg_type": "server_started", "content": {"port": port}})
+
+            # Serve until stopped
+            async with self._server:
+                await self._server.serve_forever()
+
+        # Run the server
+        try:
+            self._loop.run_until_complete(start_server())
+        except (KeyboardInterrupt, SystemExit):
+            pass
+        except asyncio.CancelledError:
+            logger.debug("Server was cancelled")
+        finally:
+            self._shutdown_server()
+
+    def start(self, lsp_host: str, shell: PositronShell, comm: BaseComm) -> None:
+        """
+        Start the LSP server.
+
+        Parameters
+        ----------
+        lsp_host
+            Host address to bind to
+        shell
+            Reference to the IPython shell for namespace access
+        comm
+            Comm for communicating with the frontend
+        """
+        self._comm = comm
+        self.shell = shell
+
+        # Reset shutdown flag if restarting
+        self.protocol._shutdown = False  # noqa: SLF001
+
+        # Stop any existing server thread
+        if self._server_thread is not None and self._server_thread.is_alive():
+            logger.warning("LSP server thread already exists, shutting it down")
+            if self._stop_event:
+                self._stop_event.set()
+                self._server_thread.join(timeout=5)
+                if self._server_thread.is_alive():
+                    logger.warning("LSP server thread did not exit after 5s, dropping it")
+
+        # Start the server in a new thread
+        logger.info("Starting LSP server thread")
+        self._server_thread = threading.Thread(
+            target=self.start_tcp,
+            args=(lsp_host,),
+            name="PositronLSPThread",
+            daemon=True,  # Allow process to exit while thread is running
+        )
+        self._server_thread.start()
+
+    def _shutdown_server(self) -> None:
```

Comment:
Might also not need all of this in latest pygls.

Nit: Thinking of it some more, it might be more flexible/readable to extract whatever we need here that is about running a pygls server in a separate thread to a subclass like `ThreadLanguageServer`